client/consensus/common/src/parachain_consensus.rs

```diff
@@ -41,64 +43,130 @@ const LOG_TARGET: &str = "cumulus-consensus";
 const RECOVERY_DELAY: RecoveryDelay =
 	RecoveryDelay { min: Duration::ZERO, max: Duration::from_secs(30) };
 
+fn handle_new_finalized_head<P, Block, B>(
+	parachain: &Arc<P>,
+	finalized_head: Vec<u8>,
+	last_seen_finalized_hashes: &mut LruMap<Block::Hash, ()>,
+) where
+	Block: BlockT,
+	B: Backend<Block>,
+	P: Finalizer<Block, B> + UsageProvider<Block> + BlockchainEvents<Block>,
+{
+	let header = match Block::Header::decode(&mut &finalized_head[..]) {
+		Ok(header) => header,
+		Err(err) => {
+			tracing::debug!(
+				target: LOG_TARGET,
+				error = ?err,
+				"Could not decode parachain header while following finalized heads.",
+			);
+			return
+		},
+	};
+
+	let hash = header.hash();
+
+	last_seen_finalized_hashes.insert(hash, ());
+	// don't finalize the same block multiple times.
+	if parachain.usage_info().chain.finalized_hash != hash {
```

Comment:
under which conditions we might attempt to refinalize the same block?

Reply:
Basically, this piece of code follows the relay chain finalization notifications. When we see one, we find the currently included parachain header for that block and finalize it. However, if the parachain did not make progress (for whatever reason) during two finalization notifications of the relay chain, the included parachain header would be the same, and we need to guard against this.